internal/osutil/permissions.go

```diff
@@ -0,0 +1,42 @@
+// Copyright (c) 2022 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package osutil
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+)
+
+// ParsePermissions parses an octal string representing Unix file permissions
+// into a os.FileMode value. If permissions is an empty string, defaultMode will
+// be used instead.
+func ParsePermissions(permissions string, defaultMode os.FileMode) (os.FileMode, error) {
+	if permissions == "" {
+		return defaultMode, nil
+	}
+
+	// Allow 777, 0777 or 1777, but never 77 or 1
+	if len(permissions) != 3 && len(permissions) != 4 {
```

Comment:
Major red-flag here: this PR is about adding mkdir functionality on the client, yet it silently moved this function while changing its behavior, and this is not even documented in the PR description.  So this is silently modifying the API of Pebble in a way that is very easy to miss. The change in behavior needs to be reverted from this PR, and moved into its own PR with an explanation about why it's required and potential consequences for Pebble usage.

To be extra clear, moving the function to osutil is fine, and I'm not saying the change in behavior is wrong. I haven't even stopped to think about it. I'm just highlighting that we need to be careful with such silent external API changes as part of unrelated PRs. Once an API changes, typically there's no going back.

In addition, this particular change also seems to disagree with what's documented elsewhere in the PR.